pex/ziputils.py

```diff
@@ -242,3 +247,93 @@ def isolate_zip(self, out_fp):
+        arcname = re.sub(r"/*$", "/", name)
+        # Unlike PermPreservingZipFile#zip_entry_from_file(), this should never correspond to a file
+        # on disk, as this class is used to synthesize zip files from cache and the created
+        # directories are also virtual. Giving it a non-zero timestamp would be misleading.
+        zinfo = zipfile.ZipInfo(filename=arcname, date_time=DETERMINISTIC_DATETIME.timetuple()[:6])
+        zinfo.file_size = 0
+        zinfo.external_attr = ((0o40000 | mode) & 0xFFFF) << 16
+        zinfo.external_attr |= 0x10
+        zinfo.compress_type = zipfile.ZIP_STORED
+        self.writestr(zinfo, b"")
+
+    # This exists to placate mypy.
+    def __enter__(self):
+        # type: () -> MergeableZipFile
+        return self
+
+    def merge_archive(self, source_zf, name_prefix=None):
+        # type: (zipfile.ZipFile, Optional[str]) -> None
+        """Copy entries from `source_path` to `destination` without decompressing.
+
+        If provided, `name_prefix` will be applied to the names of file and directory entries.
+
+        NB: This method is not aware of data descriptors, and will not copy over their contents!"""
+        assert self.fp is not None
+        assert isinstance(self.fp, io.BufferedIOBase)  # type: ignore[unreachable]
+        assert source_zf.fp is not None  # type: ignore[unreachable]
+        assert isinstance(source_zf.fp, io.BufferedIOBase)
+
+        for zinfo in source_zf.infolist():
+            # We will be mutating the ZipInfo and writing to the destination stream, so save this
+            # info upfront.
+            source_header_len = len(zinfo.FileHeader())
+            source_offset = zinfo.header_offset
+            start_of_destination_entry = self.fp.tell()
+
+            # (1) Modify the values which will affect the contents of the local file header for
+            #     this entry.
+            if name_prefix:
+                zinfo.filename = os.path.join(name_prefix, zinfo.filename)
+
+            # (2) Modify the values which will affect this entry's central directory header.
+            # The new entry will begin at the very end of the existing file.
+            zinfo.header_offset = start_of_destination_entry
+
+            # (3) Generate the modified header, then copy over the header and contents.
+            # > (3.1) Copy over the header bytes verbatim into the output zip's underlying
+            #       file handle.
+            self.fp.write(zinfo.FileHeader())
+
+            # > (3.2) Seek to the start of the source entry's file contents.
+            source_zf.fp.seek(source_offset + source_header_len, io.SEEK_SET)
+            # > (3.3) Copy over the file data verbatim from the source zip's underlying
+            #       file handle.
+            copy_file_range(source_zf.fp, self.fp, zinfo.compress_size)
+
+            # (4) Hack the synthesized ZipInfo onto the destination zip's infos.
+            self.filelist.append(zinfo)
```

Comment:
So the hack tally is `self.{fp,filelist,NameToInfo}` afaict. Techincally "public", but certainly not documented as such. CI proves this is OK across CPython 2.7, 3.5, 3.7, 3.11 and 3.12 as well as PyPy 2.7 and 3.9. CI will continue to be extensive; so I'm not too worried in the medium term, but it does seem like this should all be behind a flag given the combination of the hackyness and the new cache burden (which is you thankfully eliminated for `--layout packed` users like Pants, but still exists for folks who never use(d) that option). You seemed to be on board with a feature flag ealier - does that still seem reasonable?

Ideally the original tack of a library that handled the zipping, possibly including parallelization in addition to the ability here of merging would take the place of all this to both eliminate the hack factor and speed things up in the cold cache case not to mention make these features available outside Pex and outside Python.

Reply:
Perhaps no flag is just fine. I'm a bit conflicted here. I still do not see the real use case for this feature, I'd personally use `--layout packed` for both iteration and distribution (via rsync over ssh), but Pex has never treated new caches as opt-in. Its cache behavior is currently uniformly untunable.

Reply:
> You seemed to be on board with a feature flag ealier - does that still seem reasonable?

Yes, absolutely! Pip still has the zip file magic I proposed to them behind `--use-feature=fast-deps` for similar reasons, although that's had a lot more people contribute to it since then.

> the new cache burden (which is you thankfully eliminated for --layout packed users like Pants, but still exists for folks who never use(d) that option).

Yes: for this reason alone I think it should be behind a feature flag.

> Ideally the original tack of a library that handled the zipping, possibly including parallelization in addition to the ability here of merging would take the place of all this to both eliminate the hack factor and speed things up in the cold cache case not to mention make these features available outside Pex and outside Python.

Well yes, that definitely hasn't evaporated; in fact, it's quite useful, and indeed takes advantage of rust's structured parallelism, and it's a CLI, a rust library, and a pyo3 python library: https://github.com/cosmicexplorer/medusa-zip. I also got the hang of using github actions and cibuildwheels in the meantime and it's now published to crates.io (`{,lib,py}medusa-zip`) and pypi (`medusa-zip`) upon every tagged release. I've been continuing to speak with @jonjohnsonjr regarding the similarity of this work to Chainguard's use case efficiently merging tarballs and/or containers (see chainguard-dev/apko#781) and considering whether we can share any code. `libarchive` exists, but its goal is compatibility with the widest range of formats, not performance, and it seems plausible that a general framework for efficiently merging the contents of archives might have applicability to a wide range of build tools and package managers.

Reply:
>  outside Python.

I might propose this `ZipFile#merge_archive()` method to `python-ideas` too, although it seems possibly too specialized to be in the stdlib.

Reply:
Yeah, that seems too specialized. FWICT they have been on a somewhat lazy warpath to do less in the stdlib, not more.